class UserWindow(QDialog):
    def __init__(self, conn, cursor, user_role):
        super().__init__()
        self.conn = conn
        self.cursor = cursor
        self.user_role = user_role  # üëà –¥–æ–±–∞–≤–ª–µ–Ω–æ

        loadUi("design.ui", self)
        self.setFixedSize(self.widget.size())

        self.sport_combo = self.findChild(type(self.comboBox), "comboBox")
        self.date_edit = self.findChild(type(self.dateEdit), "dateEdit")
        self.result_button = self.findChild(type(self.pushButton), "pushButton")
        self.add_match_button = self.findChild(type(self.pushButton_2), "pushButton_2")
        self.table_view = self.findChild(type(self.tableView), "tableView")

        self.result_button.clicked.connect(self.load_match_results)
        self.add_match_button.clicked.connect(self.add_match)

        self.load_sport_types()

        self.check_permissions()  # –¥–æ–±–∞–≤–∏–º –º–µ—Ç–æ–¥ –ø—Ä–æ–≤–µ—Ä–∫–∏ –ø—Ä–∞–≤


    def load_sport_types(self):
        try:
            self.cursor.execute("SELECT SportName FROM SportType;")
            sports = self.cursor.fetchall()
            self.sport_combo.clear()
            for sport in sports:
                self.sport_combo.addItem(sport[0])
        except Exception as e:
            QMessageBox.critical(self, "–û—à–∏–±–∫–∞", f"–û—à–∏–±–∫–∞ –∑–∞–≥—Ä—É–∑–∫–∏ –≤–∏–¥–æ–≤ —Å–ø–æ—Ä—Ç–∞:\n{e}")

    def load_match_results(self):
        selected_sport = self.sport_combo.currentText()
        selected_date = self.date_edit.date().toString("yyyy-MM-dd")

        try:
            self.cursor.execute("""
                SELECT m.MatchDateTime, t1.TeamName, t2.TeamName, r1.Score, r2.Score
                FROM Match m
                JOIN Team t1 ON m.ID_Team1 = t1.ID_Team
                JOIN Team t2 ON m.ID_Team2 = t2.ID_Team
                JOIN SportType st ON m.ID_SportType = st.ID_SportType
                JOIN Result r1 ON r1.ID_Match = m.ID_Match AND r1.ID_Team = t1.ID_Team
                JOIN Result r2 ON r2.ID_Match = m.ID_Match AND r2.ID_Team = t2.ID_Team
                WHERE st.SportName = %s AND DATE(m.MatchDateTime) = %s;
            """, (selected_sport, selected_date))

            data = self.cursor.fetchall()

            model = QStandardItemModel()
            model.setHorizontalHeaderLabels(["–î–∞—Ç–∞", "–ö–æ–º–∞–Ω–¥–∞ 1", "–°—á—ë—Ç 1", "–°—á—ë—Ç 2", "–ö–æ–º–∞–Ω–¥–∞ 2"])

            for row in data:
                match_date, team1, team2, score1, score2 = row
                items = [
                    QStandardItem(str(match_date)),
                    QStandardItem(team1),
                    QStandardItem(str(score1)),
                    QStandardItem(str(score2)),
                    QStandardItem(team2),
                ]
                model.appendRow(items)
            self.table_view.setModel(model)

        except Exception as e:
            QMessageBox.critical(self, "–û—à–∏–±–∫–∞", f"–û—à–∏–±–∫–∞ –∑–∞–≥—Ä—É–∑–∫–∏ —Ä–µ–∑—É–ª—å—Ç–∞—Ç–æ–≤:\n{e}")

    def add_match(self):
        self.add_match_button = AddResult(self.conn, self.cursor)
        self.user_window.show()


class AddResult(QDialog):
    def __init__(self, conn, cursor):
        super().__init__()
        self.conn = conn
        self.cursor = cursor

        loadUi("addmatch.ui", self)  # –ù–∞–∑–≤–∞–Ω–∏–µ —Ç–≤–æ–µ–≥–æ .ui-—Ñ–∞–π–ª–∞

        self.setFixedSize(self.widget.size())

        # –ü—Ä–∏–≤—è–∑–∫–∞ –≤–∏–¥–∂–µ—Ç–æ–≤
        self.sport_combo = self.findChild(type(self.comboBox), "comboBox")
        self.date_edit = self.findChild(type(self.dateEdit), "dateEdit")
        self.team1_line = self.findChild(type(self.lineEdit), "lineEdit")
        self.team2_line = self.findChild(type(self.lineEdit_2), "lineEdit_2")
        self.score_line = self.findChild(type(self.lineEdit_3), "lineEdit_3")
        self.add_button = self.findChild(type(self.pushButton_2), "pushButton_2")

        # –ü–æ–¥–∫–ª—é—á–µ–Ω–∏–µ –∫–Ω–æ–ø–∫–∏
        self.add_button.clicked.connect(self.save_result)

        # –ó–∞–≥—Ä—É–∑–∫–∞ –≤–∏–¥–æ–≤ —Å–ø–æ—Ä—Ç–∞
        self.load_sports()

    def load_sports(self):
        try:
            self.cursor.execute("SELECT SportName FROM SportType;")
            sports = self.cursor.fetchall()
            self.sport_combo.clear()
            for sport in sports:
                self.sport_combo.addItem(sport[0])
        except Exception as e:
            QMessageBox.critical(self, "–û—à–∏–±–∫–∞", f"–û—à–∏–±–∫–∞ –∑–∞–≥—Ä—É–∑–∫–∏ –≤–∏–¥–æ–≤ —Å–ø–æ—Ä—Ç–∞:\n{e}")

    def save_result(self):
        try:
            sport_name = self.sport_combo.currentText()
            match_date = self.date_edit.date().toString("yyyy-MM-dd")
            team1_name = self.team1_line.text().strip()
            team2_name = self.team2_line.text().strip()
            score_text = self.score_line.text().strip()

            if not (team1_name and team2_name and score_text):
                QMessageBox.warning(self, "–ü—Ä–æ–≤–µ—Ä–∫–∞", "–ü–æ–∂–∞–ª—É–π—Å—Ç–∞, –∑–∞–ø–æ–ª–Ω–∏—Ç–µ –≤—Å–µ –ø–æ–ª—è.")
                return

            scores = score_text.split()
            if len(scores) != 2 or not all(s.isdigit() for s in scores):
                QMessageBox.warning(self, "–ù–µ–≤–µ—Ä–Ω—ã–π —Ñ–æ—Ä–º–∞—Ç", "–í–≤–µ–¥–∏—Ç–µ —Å—á—ë—Ç –≤ —Ñ–æ—Ä–º–∞—Ç–µ: `1 2` (—á–µ—Ä–µ–∑ –ø—Ä–æ–±–µ–ª).")
                return

            score1, score2 = int(scores[0]), int(scores[1])

            # –ü–æ–ª—É—á–µ–Ω–∏–µ ID –≤–∏–¥–∞ —Å–ø–æ—Ä—Ç–∞
            self.cursor.execute("SELECT ID_SportType FROM SportType WHERE SportName = %s;", (sport_name,))
            sport_id = self.cursor.fetchone()
            if not sport_id:
                QMessageBox.critical(self, "–û—à–∏–±–∫–∞", "–í—ã–±—Ä–∞–Ω–Ω—ã–π –≤–∏–¥ —Å–ø–æ—Ä—Ç–∞ –Ω–µ –Ω–∞–π–¥–µ–Ω.")
                return
            sport_id = sport_id[0]

            # –í—Å—Ç–∞–≤–∫–∞ –∫–æ–º–∞–Ω–¥ (–µ—Å–ª–∏ –Ω–µ —Å—É—â–µ—Å—Ç–≤—É–µ—Ç, —Ç–æ –¥–æ–±–∞–≤–ª—è–µ–º)
            self.cursor.execute("SELECT ID_Team FROM Team WHERE TeamName = %s;", (team1_name,))
            result = self.cursor.fetchone()
            if result:
                team1_id = result[0]
            else:
                self.cursor.execute("INSERT INTO Team (TeamName) VALUES (%s) RETURNING ID_Team;", (team1_name,))
                team1_id = self.cursor.fetchone()[0]

            self.cursor.execute("SELECT ID_Team FROM Team WHERE TeamName = %s;", (team2_name,))
            result = self.cursor.fetchone()
            if result:
                team2_id = result[0]
            else:
                self.cursor.execute("INSERT INTO Team (TeamName) VALUES (%s) RETURNING ID_Team;", (team2_name,))
                team2_id = self.cursor.fetchone()[0]

            # –í—Å—Ç–∞–≤–∫–∞ –º–∞—Ç—á–∞
            self.cursor.execute("""
                INSERT INTO Match (MatchDateTime, Location, ID_Team1, ID_Team2, ID_SportType)
                VALUES (%s, '–ù–µ —É–∫–∞–∑–∞–Ω–æ', %s, %s, %s)
                RETURNING ID_Match;
            """, (match_date, team1_id, team2_id, sport_id))
            match_id = self.cursor.fetchone()[0]

            # –í—Å—Ç–∞–≤–∫–∞ —Ä–µ–∑—É–ª—å—Ç–∞—Ç–æ–≤
            self.cursor.execute("""
                INSERT INTO Result (ID_Match, ID_Team, Score)
                VALUES (%s, %s, %s), (%s, %s, %s);
            """, (match_id, team1_id, score1, match_id, team2_id, score2))

            self.conn.commit()
            QMessageBox.information(self, "–£—Å–ø–µ—à–Ω–æ", "–ú–∞—Ç—á –∏ —Ä–µ–∑—É–ª—å—Ç–∞—Ç –¥–æ–±–∞–≤–ª–µ–Ω—ã!")

            # –û—á–∏—Å—Ç–∫–∞ –ø–æ–ª–µ–π
            self.team1_line.clear()
            self.team2_line.clear()
            self.score_line.clear()

        except Exception as e:
            self.conn.rollback()
            QMessageBox.critical(self, "–û—à–∏–±–∫–∞", f"–ù–µ —É–¥–∞–ª–æ—Å—å —Å–æ—Ö—Ä–∞–Ω–∏—Ç—å —Ä–µ–∑—É–ª—å—Ç–∞—Ç:\n{e}")
